### PR TITLE
Add telemetry to callback service

### PR DESCRIPTION
**Pull request by dtivel**
Manually verified end-to-end experience and telemetry in AppInsights.

Also:

* ignore tools\nuget.exe downloaded by running build.cmd
* fix build warnings around improper use of `async` keyword.
* remove text responses from callback service (except for GET).

### CODE REVIEW

File: src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs [hunk 1]
```diff
@@ -38,10 +41,26 @@ public VcsCallbackServerStartup()
             _packageValidationTable = new PackageValidationTable(cloudStorageAccount, containerName);
             _packageValidationAuditor = new PackageValidationAuditor(cloudStorageAccount, containerName);
             _notificationService = new NotificationService(cloudStorageAccount, containerName);
+
+            // Setup telemetry
+            _instrumentationKey = configurationService.Get("AppInsightsInstrumentationKey").Result;
```
Comment (shishirx34) [inline src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs]: @scottbommarito - This might need Keyvault rotation and integration? Not sure if it is already implemented here.
`cloudStorageAccount` uses storage values.

File: src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs [hunk 2]
```diff
@@ -183,24 +216,38 @@ public void Configuration(IAppBuilder app)
                         }
                     }
                 }
-
-                // "OK"
-                context.Response.ContentType = "text/plain";
-                await context.Response.WriteAsync("Thank you, come again.");
```
Comment (shishirx34) [inline src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs]: Was this the bug? Or this path was never hit?

Comment (scottbommarito) [inline src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs]: The bug seemed like it was that they expected a `text/xml` response but received a `text/plain` response.

File: src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs (hunk 1 above)
Comment (scottbommarito) [inline src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs]: It needs secret rotation integration: `ConfigurationService` is integrated with KeyVault but given this implementation it will not refresh itself.

File: src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs (hunk 2 above)
Comment (shishirx34) [inline src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs]: I know that, hence the question, is this the relevant route? There are a couple more which return text/plain. If the route reached here, it means something didn't add up in the validations above, which is where the bug could be potentially.

Comment (scottbommarito) [inline src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs]: This is the relevant route. The "Thank you, come again" message was explicitly in the exception trace.

File: src/Validation.Common/Validators/Vcs/ApplicationInsightsMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        private static void TrackRequest(IOwinContext context, DateTimeOffset startTime, TimeSpan duration)
+        {
+            var telemetry = new RequestTelemetry()
+            {
+                Duration = duration,
+                HttpMethod = context.Request.Method,
+                Name = TelemetryConstants.Vcs,
+                ResponseCode = context.Response.StatusCode.ToString(),
+                StartTime = startTime,
+                Success = context.Response.StatusCode < 400,
```
Comment (dtivel) [inline src/Validation.Common/Validators/Vcs/ApplicationInsightsMiddleware.cs]: You're correct.  https://github.com/Microsoft/ApplicationInsights-dotnet/blob/develop/src/Core/Managed/Shared/DataContracts/RequestTelemetry.cs#L224

File: src/Tests.Logger/Tests.Logger.Job.cs [hunk 4]
```diff
@@ -32,9 +32,9 @@ internal class Job : JobBase
         public override bool Init(IDictionary<string, string> jobArgsDictionary)
         {
             JobScenario = jobArgsDictionary.GetOrNull<int>(ScenarioArgumentName);
-            if(JobScenario == null)
+            if (JobScenario == null)
             {
-                throw new ArgumentException("Argument '"+ ScenarioArgumentName +"' is mandatory." + HelpMessage);
+                throw new ArgumentException("Argument '" + ScenarioArgumentName + "' is mandatory." + HelpMessage);
```
Comment (skofman1) [inline src/Tests.Logger/Tests.Logger.Job.cs]: you can use c# 6 formatting: $"Argument '{ScenarioArgumentName}' is mandatory. {HelpMessage}"

File: src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs (hunk 1 above)
Comment (skofman1) [inline src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs]: AppInsightsInstrumentationKey is not a secret

File: src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs [hunk 5]
```diff
@@ -63,6 +82,12 @@ public void Configuration(IAppBuilder app)
                         validationEntity = await _packageValidationTable.GetValidationAsync(validationId);
                         if (validationEntity == null)
                         {
+                            var telemetry = new EventTelemetry(TelemetryConstants.RequestNotFound);
```
Comment (skofman1) [inline src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs]: You can use this logging library: https://github.com/NuGet/ServerCommon/tree/develop/src/NuGet.Services.Logging 
that is used in all server components. It uses serilog for logging. Serilog allows you to print messages in a formatted form that also adds parameters as properties in AppInsights. This way, this implementation can be much simpler.

File: src/Validation.Common/Validators/Vcs/ApplicationInsightsMiddleware.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.Owin;
+
+namespace NuGet.Jobs.Validation.Common.Validators.Vcs
+{
+    internal sealed class ApplicationInsightsMiddleware : OwinMiddleware
```
Comment (skofman1) [inline src/Validation.Common/Validators/Vcs/ApplicationInsightsMiddleware.cs]: why is this needed? Gallery is built with owin, but doesn't have a similar middleware implementation. I was assuming AI just plugs in automagically into the pipeline...

File: src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs (hunk 2 above)
Comment (scottbommarito) [inline src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs]: SOAP response was fixed in a different PR.

File: src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs (hunk 1 above)
Comment (scottbommarito) [inline src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs]: There are other secrets that are not `AppInsightsInstrumentationKey` (for example storage account credentials) that are used by the service.

Comment (scottbommarito) [inline src/Validation.Common/Validators/Vcs/VcsCallbackServer.cs]: Secret rotation integration will be handled by another PR.

File: src/Validation.Common/Validators/Vcs/ApplicationInsightsMiddleware.cs (hunk 6 above)
Comment (scottbommarito) [inline src/Validation.Common/Validators/Vcs/ApplicationInsightsMiddleware.cs]: It's much simpler in the gallery because the gallery is built with MVC. AI plugs in automagically into the MVC pipeline, but not the OWIN pipeline. [The search service, which uses OWIN but not MVC, like this project, has a very similar implementation to this one that we should use here instead of implementing it again.](https://github.com/NuGet/NuGet.Services.Metadata/tree/master/src/NuGet.ApplicationInsights.Owin)

Comment (scottbommarito) [inline src/Validation.Common/Validators/Vcs/ApplicationInsightsMiddleware.cs]: I will move the package in the search service that contains the OWIN middleware that enables AI logging to Server.Common so that we can use it here as well.

Comment (skofman1) [inline src/Validation.Common/Validators/Vcs/ApplicationInsightsMiddleware.cs]: It's already available in MyGet. This work is not required to use the package here.

Comment (scottbommarito) [inline src/Validation.Common/Validators/Vcs/ApplicationInsightsMiddleware.cs]: but ideally it should be in a shared place regardless

Comment (shishirx34) [inline src/Validation.Common/Validators/Vcs/ApplicationInsightsMiddleware.cs]: I would be against unnecessary refactoring. If it is a small amount, rather duplicate it and move on.
